### PR TITLE
Fix Canary not stripping prompt from reference + more test coverage

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

TL;DR: fix for a bug I accidentally introduced in #9901 

Context:
* #9901 changed multi-task beam search decoding module to strip the prompt and EOS away form recognitions; it removed "remove_special_tokens" sprinkled in various places in our codebase (transcribe, wer, bleu, etc.)
* That fix was actually porting a part of my changes in an old closed PR #8833. However, I forgot that we also have to handle the dataset part properly - return GT transcripts with stripped prompts and EOS (which was done there). So I ended up introduced a bug that reference has the prompt, but recognition doesn't in scoring.
* I ported the necessary additional changes from #8833 and adapted them to the current codebase.
* Since I wanted to improve the test coverage of the affected parts, I also pulled the changes from #8885 (another PR I forgot to merge) where I already wrote these tests, and extended them.

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 1]
```diff
@@ -180,8 +199,9 @@ def canary(
         )
         prompts_with_answers.append(encoded["input_ids"])
         prompts.append(encoded["context_ids"])
+        answers.append(encoded["answer_ids"][:-1])  # Strip Canary's EOS
```
Comment (bonham79) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:202]: Since the current issue comes from stripping buried int the code, let's have a bool here so we're aware that this is happening.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 2]
```diff
@@ -664,20 +664,18 @@ def forward(
...
-    def training_step(self, batch, batch_nb):
+    def training_step(self, batch: PromptedAudioToTextMiniBatch, batch_nb):
 
         if batch is None:
             return torch.tensor([0.0])
 
-        # During training prompt and prompt_len are null, ignore.
-        signal, signal_len, transcript, transcript_len, prompt, prompt_len = batch
-        input_ids, labels = transcript[:, :-1], transcript[:, 1:]
+        input_ids, labels = batch.prompted_transcript[:, :-1], batch.prompted_transcript[:, 1:]
```
Comment (bonham79) [inline nemo/collections/asr/models/aed_multitask_models.py:672]: Add doc strings for the indexing removing BOS and EOS. You could also just add flags to the minibatch method that does this for you.

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 3]
```diff
@@ -25,6 +26,18 @@
 from nemo.collections.common.tokenizers.canary_tokenizer import CANARY_SPECIAL_TOKENIZER
 
 
+@dataclass
+class PromptedAudioToTextMiniBatch:
```
Comment (bonham79) [inline nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:30]: Per @titu1994 from offthread convo.

```


PTL will not auto cast tensors tht are inside data classes, only inside simpler structures like lists, tuples and dicts.
We can try to use TypedDict collection, that might work```


[11:12 AM](https://nvidia.slack.com/archives/D02KMNFH6UC/p1721844762008059)
Just use torch.Tensor or Any as types if needed
11:13 AM
We need to test if PTL can handle the change properly and still continue to cast inputs automatically for the batch
[11:13 AM](https://nvidia.slack.com/archives/D02KMNFH6UC/p1721844800900519)
Worst comes to worse we can ask PTL to check for and cast tensors to service from inside TypedDicts though that will take time

'''

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 4]
```diff
@@ -145,30 +163,52 @@ def _forward(
...
+            encoder_hidden_states = torch.repeat_interleave(encoder_hidden_states, self.n_samples, dim=0)
+            encoder_input_mask = torch.repeat_interleave(encoder_input_mask, self.n_samples, dim=0)
+            orig_batch_size = batch_size
+            batch_size = batch_size * self.n_samples
 
         # pad profile tracks sequences ending with <eos> token to replace
         # everything after <eos> with <pad> token
         decoder_parameter = next(self.decoder.parameters())
-        pad_profile = torch.zeros(batch_size, 1).long().to(decoder_parameter.device)
+        pad_profile = torch.zeros(batch_size).long().to(decoder_parameter.device)
```
Comment (bonham79) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:179]: What's the change that allows dimension?

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 5]
```diff
@@ -145,30 +163,52 @@ def _forward(
...
 
         decoder_mems_list = None
         for i in range(max_generation_length):
 
-            log_probs, decoder_mems_list = self._one_step_forward(
-                tgt[:, -1:], encoder_hidden_states, encoder_input_mask, decoder_mems_list, i
+            if i == 0:
+                input_ids = tgt
+            else:
+                input_ids = tgt[:, -1:]
```
Comment (bonham79) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:187]: San check: Is this doubling any of the slicing you do in the aed model?

File: nemo/collections/asr/parts/submodules/multitask_decoding.py [hunk 6]
```diff
@@ -103,30 +105,43 @@ def __init__(
...
-        if self.cfg.strategy == 'greedy' or self.cfg.strategy == 'greedy_batch':
+        if strategy in ['greedy', 'greedy_batch']:
 
-            # self.decoding = None
-            raise NotImplementedError("Greedy decoding is not implemented yet.")
+            self.decoding = TransformerAEDGreedyInfer(
+                transformer_decoder=self.transformer_decoder,
+                log_softmax_module=self.log_softmax_module,
+                tokenizer=self.tokenizer,
+                max_generation_delta=self.cfg.greedy.get('max_generation_delta', 50),
```
Comment (bonham79) [inline nemo/collections/asr/parts/submodules/multitask_decoding.py:133]: Make the default 1max_generation_delta1 `-1`. I've had way too many headaches tracking down the default numbers in decoding.

File: nemo/collections/asr/parts/submodules/token_classifier.py [hunk 7]
```diff
@@ -81,8 +88,12 @@ def __init__(
         )
         self.post_init(use_transformer_init=use_transformer_init)
 
+    def set_log_softmax_enabled(self, value: bool) -> "TokenClassifier":
```
Comment (bonham79) [inline nemo/collections/asr/parts/submodules/token_classifier.py:91]: Add the setter and getter decorators for this. Also `self._log_softmax`. We break enough properties that should use the underscoring convention.

File: nemo/collections/asr/parts/submodules/token_classifier.py [hunk 8]
```diff
@@ -147,8 +164,12 @@ def __init__(
         )
         self.post_init(use_transformer_init=use_transformer_init)
 
+    def set_log_softmax_enabled(self, value: bool) -> "BertPretrainingTokenClassifier":
```
Comment (bonham79) [inline nemo/collections/asr/parts/submodules/token_classifier.py:167]: Ditto to previous

File: nemo/collections/common/parts/multi_layer_perceptron.py [hunk 9]
```diff
@@ -51,11 +51,15 @@ def __init__(
     def last_linear_layer(self):
         return getattr(self, f'layer{self.layers - 1}')
 
-    def forward(self, hidden_states):
+    def forward(self, hidden_states, temperature: float | None = None):
```
Comment (bonham79) [inline nemo/collections/common/parts/multi_layer_perceptron.py:54]: Is there another place you can add the temperature? I prefer MLP base classes to be as brain dead as possible. Even with an optional flag....

File: tests/collections/asr/decoding/test_multi_task_decoding.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+)
+from nemo.collections.asr.parts.submodules.multitask_beam_decoding import TransformerAEDBeamInfer
+from nemo.collections.asr.parts.submodules.multitask_greedy_decoding import TransformerAEDGreedyInfer
+from nemo.collections.asr.parts.submodules.token_classifier import TokenClassifier
+
+
+@pytest.fixture()
+def deterministic_rng():
+    state = torch.get_rng_state()
+    torch.manual_seed(0)
```
Comment (bonham79) [inline tests/collections/asr/decoding/test_multi_task_decoding.py:19]: I'm not up to date on torch's rng, but isn't setting the manual seed first safer?

File: tests/collections/asr/decoding/test_multi_task_decoding.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    (scores_seq0,) = scores
+    assert torch.is_tensor(scores_seq0)
+    assert scores_seq0.shape == (2,)
+
+    assert best_path is not None
+    assert torch.is_tensor(best_path)
+    assert best_path.shape == (1, 26)
+
+
+def test_transformer_aed_beam_infer_strips_prompt(inputs, decoder_nm, nnet, tokenizer):
```
Comment (bonham79) [inline tests/collections/asr/decoding/test_multi_task_decoding.py:135]: Sorry: what's the sample prompt here? I'm just seeing `BOS` + `EOS`. The test should have variable length prompts just so it can extend beyond canary cases.

**Review by bonham79 — COMMENTED** [review]
Left comments, also for sanity checking, can you run a toy model against a stable release and this branch just to ensure nothing pops up?

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py (hunk 3 above)
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:30]: It worked fine when I ran this, maybe they added dataclass support later.

File: tests/collections/asr/decoding/test_multi_task_decoding.py (hunk 10 above)
Comment (pzelasko) [reply tests/collections/asr/decoding/test_multi_task_decoding.py:19]: You need to keep a copy of the current state before you seed, so that you can restore it after the test.

File: tests/collections/asr/decoding/test_multi_task_decoding.py (hunk 11 above)
Comment (pzelasko) [reply tests/collections/asr/decoding/test_multi_task_decoding.py:135]: It's just BOS. Good point, I'll add an actual prompt.

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 5 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer/transformer_generators.py:187]: No, it's just slicing to get the last predicted token to feed to the decoder in the current step.

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 4 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer/transformer_generators.py:179]: Removed `keepdim=True` in argmax/Categorical a few lines below

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py (hunk 3 above)
Comment (bonham79) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:30]: When you were running were you checking casting at all?

Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:30]: I launched a multi-GPU training with this change and it runs fine in terms of GPU utilization, so the batch had to be moved to CUDA.

Comment (tbartley94) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py:30]: Marvelous

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 12]
```diff
@@ -75,6 +85,7 @@ def _one_step_forward(
         encoder_input_mask=None,
         decoder_mems_list=None,
         pos=0,
+        need_scores: bool = True,
```
Comment (tbartley94) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:88]: Can we change the variable to `return_scores` to be consistent with other commands like this?

File: nemo/collections/asr/parts/submodules/token_classifier.py (hunk 7 above)
Comment (tbartley94) [reply nemo/collections/asr/parts/submodules/token_classifier.py]: Ooh, good call.

**Review by tbartley94 — APPROVED** [review]
One name change, fix import flags from the CI.

Btw, how are you feeding temperature to the MLP now?

**Comment by pzelasko** [comment]
> One name change, fix import flags from the CI.

OK

> Btw, how are you feeding temperature to the MLP now?

It's applied directly in `Categorical`

**Review by tbartley94 — APPROVED** [review]
